File: app/src/main/java/io/github/zwieback/familyfinance/business/chart/service/formatter/QuarterValueFormatter.java
package io.github.zwieback.familyfinance.business.chart.service.formatter;

import com.github.mikephil.charting.formatter.ValueFormatter;

import org.threeten.bp.LocalDate;

import io.github.zwieback.familyfinance.util.DateUtils;

public class QuarterValueFormatter extends ValueFormatter {

    @Override
    public String getFormattedValue(float quartersFromEpoch) {
        LocalDate quarter = determineCorrectQuarter(quartersFromEpoch);
        return DateUtils.localDateToString(quarter, DateUtils.ISO_LOCAL_QUARTER);
    }

    private static LocalDate determineCorrectQuarter(float quartersFromEpoch) {
        return DateUtils.epochQuarterToLocalDate((long) quartersFromEpoch);
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/chart/service/formatter/QuarterValueFormatter.kt
package io.github.zwieback.familyfinance.business.chart.service.formatter

import io.github.zwieback.familyfinance.constant.DateConstants
import io.github.zwieback.familyfinance.extension.QuartersFromEpoch
import org.threeten.bp.format.DateTimeFormatter

class QuarterValueFormatter : SpecificPeriodFormatter() {

    override fun getFormatter(): DateTimeFormatter = DateConstants.ISO_LOCAL_QUARTER

    override fun getSpecificPeriod(value: Long) = QuartersFromEpoch(value)
}
